### PR TITLE
feat: GoDAM 2.0 onboarding — embedded SPA + godam-core integration

### PR DESCRIPTION
**Pull request by subodhr258**
## Summary

Implements the **GoDAM 2.0 onboarding** experience (tracking: rtCamp/godam-plugin-wp#5) — an embedded wp-admin SPA wired to the live **godam-core** auth API. It **overlays the Dashboard + Analytics screens whenever the site isn't connected**, guiding admins to connect without leaving wp-admin.

Every auth/account screen renders as a centered modal over the dimmed page, matching the Figma. The **durable credential stays the org API key** — the GoDAM JWT is transient and held server-side, so onboarding's end-state equals the current connected state (back-compat friendly).

## Architecture
- **Onboarding SPA** at `pages/onboarding/` (Redux Toolkit step-machine + RTK Query). Renders into `#root-godam-onboarding`, overlaying the Dashboard + Analytics when `! rtgodam_is_api_key_valid() && current_user_can( 'manage_options' )`. No separate menu page and no dev feature flag — onboarding is the connection path.
- **WP REST proxy** `inc/classes/rest-api/class-onboarding.php` (`/godam/v1/onboarding/*`) forwards to `godam_core.api.login.*` + `organization_access.*`. The SPA never calls godam-core directly. (This class also hosts the Video Editor product-guide state + the Woo nudge state — see below.)
- **JWT held server-side** (per-user transient, 24h) — never exposed to JS. Used only to `list_my_organizations` + `get_organization_api_key`, then discarded.
- **Connect** (workspace-select & license-key) goes through the existing `rtgodam_verify_api_key()` (stores the org key + registers the site).
- **Google** = **popup + `postMessage`** flow: `get_oauth2_url` (with `wordpress_site`) opens a popup → godam-core's completion page `postMessage`s a one-time code back to the SPA (origin-pinned to the GoDAM app host) → `exchange_oauth_code` exchanges it **server-side** for the JWT. The browser never sees the JWT.
- Follows the site's `RTGODAM_API_BASE` (filter `rtgodam_onboarding_api_base`).

## Screens
Entry · Signup (Google-first + verify-email gate) · Login · Forgot password · License key · Google + workspace select (plan badge + role + "0 workspaces" variant + "can't switch" notice) · Welcome. All use the shared modal shell + two side-panel variants (feature grid / social).

## Connected-state surfaces (O8 / O11)
- **O8 trial-countdown banner** (`TrialCountdownBanner`) — full-width purple "Free trial ends on `<date>`" + "Upgrade Now", shown across GoDAM pages while on the free plan. Drives off the Free subscription's `end_date` (godam-core models the 60-day trial as a Free subscription, so `trial_end_date` is always null).
- **O11 upgrade confirmation** (`ProUpgradeNotice`) — the "Upgrade plan" CTA opens the `/godam_upgrade` round-trip (org known) or the billing page (fallback); on return, the "You are now a pro member" toast shows **only once the refreshed plan is actually paid** (the `?godam_upgrade=success` param alone never triggers it).
- The old in-dashboard / in-analytics **api-key overlay was removed** (the onboarding overlay now owns the disconnected case). A focused **"analytics unavailable" notice** replaces the old overlay's connected-backend-error message. The connected dashboard visual itself is the merged #1969 analytics reskin.

## WooCommerce (O9 / O10)
- **O9 — Woo signup variant:** when WooCommerce is active (`isWoo` localized), the Entry screen shows the Woo-specific heading/intro ("You've Unlocked Woo-Specific GoDAM" + WooCommerce copy) with the feature side panel, and the connected Video Editor empty state adds "You can start selling as soon as you add a product."
- **O10 — Woo post-install nudge:** a one-time "You've Unlocked Woo Features" modal (`WooUnlockedNotice`) shown once per user when WooCommerce is active on a connected site; dismissal is persisted via the `onboarding/woo-nudge` REST route (user meta). The pro-features preview panel is an intentional placeholder for now.

## Design fidelity (Figma)
Built to the Figma tokens: accent `#5D31FF`, lavender panel `#F2F2FF`, chips `#E8E7FF`, text `#020617`/`#64748B`, radii (modal 16 / primary CTA 2 / inputs 4 / chip 8), 40px controls, type scale (entry 32/40, form H1 28/32, labels 13). Brand accent is a fixed GoDAM purple (not the admin colour scheme), per the design. **Geist** is bundled (`@fontsource/geist`).

## Testing
- **JS unit:** `pages/onboarding/utils/validators.test.js` — `npm run test:unit`.
- **PHP unit:** `tests/php/OnboardingResponseTest.php` (Frappe message unwrap + nested / `_server_messages` error extraction, incl. HTML-strip + malformed-JSON edge cases) — `composer phpunit`.
- **Manual (Chrome):** entry/signup/login/license/workspace render-verified; trial banner + "pro member" toast verified; Google `get_oauth2_url` returns a valid authorize URL; no console errors.
- Build, ESLint, stylelint, PHPCS all clean.

## How to test
On a **disconnected** site (no valid key), open **GoDAM → Dashboard** (or Analytics) — the onboarding overlay appears automatically; click through signup / login / license. To test the **trial banner + upgrade toast**, use a connected account on the free plan (a future `end_date`). To test **O9/O10**, activate **WooCommerce**: the entry screen switches to the Woo variant, and the "unlocked Woo features" nudge shows once on a connected site.

## Automation impact
Interactive elements carry `data-test-id` (`godam-<area>-<type>-<name>`). `GODAM_E2E` is localized **and now consumed** — `useGoogleSignIn` short-circuits the popup/OAuth with a deterministic code under `GODAM_E2E`, so Google sign-in is E2E-drivable (godam-core must accept the mock code in E2E mode). New test ids: `godam-header-notice-woo-unlocked`, `godam-header-button-woo-get-started`. **Existing flows change:** the old api-key overlay on Dashboard/Analytics is replaced by the onboarding overlay when disconnected.

## Known follow-ups
- The Entry "social" side panel and the O10 pro-features panel are **intentional placeholder areas** (graphics to be added later — kept empty by design).
- **"Keep me signed in"** was **removed** for now — it had no backend (godam-core `password_login` ignores `remember`, and the onboarding JWT is discarded after connect). Re-add once a persistent-session backend exists.
- **Tests** — jest for the screens + `useGoogleSignIn` + `TrialCountdownBanner` + `ProUpgradeNotice`, and PHPUnit for the REST proxy endpoints + `should_overlay_onboarding()`, need the **wp-phpunit harness** (the current bootstrap is pure-no-WP by design).
- **`data-test-id`** on the GoDAM header controls (help / Manage Media / Upgrade plan / Get GoDAM) — deferred.
- **Copy** — the SidePanel marketing strings are still placeholder pending final copy.
- **godam-core deps.** ✅ **Done:** (1) `verify_api_key` now returns `organization`, consumed directly via `userData.userApiData.organization` for every connected site incl. license-key (the interim `rtgodam_organization` option was removed); (2) the per-`site_url` verify cache is now flushed on upgrade, so the new plan reflects immediately. ⏳ **Still pending:** (3) **forgot-password:** `godam_core.api.user.reset_password` requires an `api_key` once a real external user is found (verified by probe: a fake user returns "User not found" without an api_key, a real user returns "API key is required."), but onboarding sites are disconnected and have none, so it should scope external resets via `wordpress_site`/origin (like `signup`) or drop the requirement. Plus the Google completion page (`exchange_oauth_code`) live on the target env.

### Screenshots and Recordings
Screenshot 2026-06-29 at 2 56 57 PM
Screenshot 2026-06-29 at 2 56 27 PM
Screenshot 2026-06-29 at 2 56 17 PM

https://github.com/user-attachments/assets/25b4c16a-ee6e-456e-bbda-1ed09133dd89

Note: The free-trial cache fix and the Organisation name in API Key verification are now done on GoDAM-core (the upgrade flow consumes the org name directly). Access control is still pending.

### CODE REVIEW

File: inc/classes/rest-api/class-onboarding.php [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+			'timeout' => 15, // phpcs:ignore WordPressVIPMinimum.Performance.RemoteRequestTimeout.timeout_timeout
+			'headers' => $headers,
+		);
+
+		if ( 'GET' === $method ) {
+			$response = wp_remote_get( add_query_arg( array_map( 'rawurlencode', $args ), $url ), $request_args );
+		} else {
+			$request_args['body'] = $args; // Form-encoded — matches the godam-core API.
+			$response             = wp_remote_post( $url, $request_args );
+		}
```
[the remark this answers is not in the document]
Comment (subodhr258) [reply inc/classes/rest-api/class-onboarding.php:218]: Addressed in the current code — `call_core()` no longer pre-encodes GET args; they're passed directly to `add_query_arg()` (which URL-encodes once), so there's no double-encoding. The GET branch carries an explicit comment to that effect.

File: inc/classes/class-pages.php [hunk 2]
```diff
@@ -402,6 +405,11 @@ public function render_dashboard_page() {
 			<div id="root-video-dashboard" class="<?php echo $is_premium_user ? '' : 'free-user'; ?>"></div>
 		</div>
 		<?php
+		// When the site isn't connected, overlay the 2.0 onboarding modal on the
+		// (empty-state) dashboard that renders behind it.
+		if ( ! $is_premium_user && $this->is_onboarding_v2_enabled() ) {
+			echo '<div id="root-godam-onboarding"></div>';
+		}
```
[the remark this answers is not in the document]
Comment (subodhr258) [reply inc/classes/class-pages.php:412]: Addressed — the onboarding overlay only renders when `should_overlay_onboarding()` returns true, which requires `current_user_can( 'manage_options' )`. An editor without that capability gets neither the overlay nor the enqueued bundle, so the UI and REST permissions are aligned. (The feature-flagged page this referred to has since been removed — onboarding now overlays the Dashboard/Analytics.)

File: pages/onboarding/components/screens/SignupScreen.jsx [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+				<TextControl __nextHasNoMarginBottom type="password" label={ <>{ __( 'Confirm Password', 'godam' ) } <Required /></> } value={ fields.confirm } onChange={ set( 'confirm' ) } help={ errors.confirm } placeholder={ __( 'Enter Password', 'godam' ) } data-test-id="godam-onboarding-input-confirm" />
+
+				<CheckboxControl __nextHasNoMarginBottom checked={ fields.tnc } onChange={ set( 'tnc' ) } help={ errors.tnc } data-test-id="godam-onboarding-checkbox-tnc"
+					label={ <>{ __( 'I agree to', 'godam' ) } <a className="godam-onboarding__link" href="https://godam.io/terms/" target="_blank" rel="noreferrer">{ __( 'Terms and Conditions, Privacy Policy, Refund Policy', 'godam' ) }</a></> } />
+				<CheckboxControl __nextHasNoMarginBottom checked={ fields.newsletter } onChange={ set( 'newsletter' ) } label={ __( 'I agree to Subscribe Newsletter to receive future updates and releases news.', 'godam' ) } data-test-id="godam-onboarding-checkbox-newsletter" />
+			</div>
```
[the remark this answers is not in the document]
Comment (subodhr258) [reply pages/onboarding/components/screens/SignupScreen.jsx:84]: Addressed — the newsletter label now reads: "Email me product updates and release news."

File: pages/onboarding/components/screens/LicenseKeyScreen.jsx [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+/**
+ * O5 — "Activate with license key" (reuses the existing verify_api_key flow;
+ * not a JWT login). On success the plugin connects directly.
+ */
+const LicenseKeyScreen = () => {
+	const dispatch = useDispatch();
+	const [ verifyLicenseKey, { isLoading } ] = useVerifyLicenseKeyMutation();
+	const [ key, setKey ] = useState( '' );
+	const [ remember, setRemember ] = useState( true );
+
```
[the remark this answers is not in the document]
Comment (subodhr258) [reply pages/onboarding/components/screens/LicenseKeyScreen.jsx:33]: Addressed — the License screen was rewritten and no longer has the `remember` state/checkbox. (The Login screen keeps "Keep me signed in" as a cosmetic control by product decision; wiring it server-side — godam-core has no `remember` param today — is a tracked follow-up.)

File: inc/classes/rest-api/class-onboarding-response.php [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+			return $inner;
+		}
+
+		if ( ! empty( $body['_server_messages'] ) ) {
+			$messages = json_decode( (string) $body['_server_messages'], true );
+			if ( is_array( $messages ) && ! empty( $messages[0] ) ) {
+				$first = json_decode( (string) $messages[0], true );
+				if ( isset( $first['message'] ) && is_string( $first['message'] ) ) {
+					return trim( wp_strip_all_tags( $first['message'] ) );
+				}
```
[the remark this answers is not in the document]
Comment (subodhr258) [reply inc/classes/rest-api/class-onboarding-response.php:67]: Addressed — `error_message()` now guards with `function_exists( 'wp_strip_all_tags' )` and falls back to `strip_tags()`, so `Onboarding_Response` stays framework-free (usable without WordPress).

File: pages/godam/components/ProUpgradeNotice.jsx [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+		// Second pass (post-reload): plan is fresh → confirm, then strip the query
+		// so a later reload doesn't re-trigger.
+		window.sessionStorage.removeItem( REFRESHED_KEY );
+		params.delete( 'godam_upgrade' );
+		params.delete( 'plan' );
+		params.delete( 'organization' );
+		const qs = params.toString();
+		window.history.replaceState( {}, '', window.location.pathname + ( qs ? `?${ qs }` : '' ) );
+		setShow( true );
```
[the remark this answers is not in the document]
Comment (subodhr258) [reply pages/godam/components/ProUpgradeNotice.jsx]: Fixed in ca8bd3c9 — the toast now shows only when the refreshed plan is actually paid (validApiKey && active_plan !== 'free'), after the query is stripped, so a hand-edited or stale `?godam_upgrade=success` no longer triggers a false confirmation.

File: pages/onboarding/components/OnboardingModal.jsx [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+ * @param {string}      [props.sidePanel] Side-panel variant for split layout.
+ */
+const OnboardingModal = ( { children, layout = 'split', sidePanel = 'features' } ) => {
+	const notice = useSelector( ( state ) => state.onboarding.notice );
+	const dispatch = useDispatch();
+
+	return (
+		<div className="godam-onboarding" data-test-id="godam-onboarding-root">
+			<div className="godam-onboarding__scrim" />
+			<div className={ `godam-onboarding__modal godam-onboarding__modal--${ layout }` } role="dialog" aria-modal="true">
```
[the remark this answers is not in the document]
Comment (subodhr258) [reply pages/onboarding/components/OnboardingModal.jsx]: Fixed in ca8bd3c9 — added an accessible name (aria-label) on the dialog container.

File: pages/onboarding/utils/validators.js [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+ * @param {boolean} fields.tnc       Terms accepted.
+ * @return {Object} Map of field name → error message.
+ */
+export const validateSignup = ( { firstName, email, password, confirm, tnc } ) => {
+	const errors = {};
+	if ( ! firstName || ! firstName.trim() ) {
+		errors.firstName = 'First name is required.';
+	}
+	if ( ! isValidEmail( email ) ) {
+		errors.email = 'Enter a valid email address.';
```
[the remark this answers is not in the document]
Comment (subodhr258) [reply pages/onboarding/utils/validators.js]: Fixed in ca8bd3c9 — the validator error strings are wrapped in __() so they're extracted for translation (the file still imports only @wordpress/i18n, no React).

**Comment by KMchaudhary** [comment]
## Code Review: GoDAM 2.0 onboarding — embedded SPA + godam-core integration

Reviewed the full diff (45 files). This is a well-architected change — the server-side JWT, the reuse of `rtgodam_verify_api_key()` as the durable end-state, the pure/testable response parser, and the spoof-resistant upgrade confirmation are all done thoughtfully. Findings below are mostly hardening/polish; nothing blocks merge.

### Findings

| # | File | Issue | Severity |
|---|------|-------|----------|
| 1 | `class-onboarding.php` `call_core()` | Credentials (signup/login `password`) are POSTed to `api_base()`. If `RTGODAM_API_BASE` or the `rtgodam_onboarding_api_base` filter ever resolves to `http://`, passwords/JWT travel in cleartext. Consider rejecting a non-`https` scheme in `call_core()` (allowing `http` only for `localhost`). | 🟠 Medium |
| 2 | `class-onboarding.php` `check_user_exists()` | `email` param is passed through unsanitized, unlike every other handler (`sanitize_email()`). It's URL-encoded by `add_query_arg`, so low risk, but inconsistent. Also: this route + `useCheckUserExistsMutation` appear **unused** — the entry screen routes to signup/login by user choice, never auto-branches. Either wire it up or drop it. | 🟡 Minor |
| 3 | `class-onboarding.php` `reset_password()`/`resend_verification()`/`signup()` | These proxy godam-core's email-sending endpoints with no local throttle. Admin-gated, so the blast radius is small, but a compromised/curious admin can drive reset/verification emails to arbitrary addresses through the site. Worth a brief rate-limit or an explicit note that godam-core throttles these. | 🟡 Minor |
| 4 | `class-onboarding.php` `handle_token_response()` | `set_transient(..., (int) $result['expires_in'])` trusts the server's TTL uncapped. Consider `min( $ttl, DAY_IN_SECONDS )` so a bad/huge value can't pin a stale JWT server-side. | 🟡 Minor |
| 5 | `WooUnlockedNotice.jsx` | The scrim's `onClick={ dismiss }` **persists** `seen` via the POST. An accidental backdrop click permanently kills this one-time nudge. Suggest closing-without-persisting on scrim click, and only persisting on the explicit "Get Started" CTA. | 🟡 Minor |
| 6 | `OnboardingModal.jsx` | `role="dialog" aria-modal="true"` with no focus trap or Esc handling. Keyboard/SR users can tab into the dimmed page behind. a11y gap on the primary connection path. | 🟡 Minor |
| 7 | `ProUpgradeNotice.jsx` | The post-return re-verify uses `window.wpApiSettings?.nonce`, but the onboarding flow localizes its nonce under `godamOnboarding.nonce`. On a connected page `wpApiSettings` is usually present, but if it isn't the refresh silently fails (`|| ''`) and the "pro member" toast never shows. Worth falling back to the same nonce source the rest of the code uses. | 🟡 Minor |
| 8 | `dashboard/index.scss` + `analytics/index.scss` | Identical `.godam-analytics-unavailable { margin: 16px 0; }` duplicated in both. Since the component is shared (`pages/shared/`), the style could live once alongside it. | 🟢 Nit |

### What looks good
- **Google popup handoff** — origin-pinned (`event.origin !== config.appOrigin`) *and* shape-checked (`source !== 'godam-oauth'`), popup opened synchronously on click, cancellation polling, fail-fast when `appOrigin` is missing. Solid.
- **JWT never reaches the browser** — held in a per-user transient, discarded after connect.
- **Spoof-resistant upgrade toast** — gating on the refreshed *paid* plan rather than the `?godam_upgrade=success` param is exactly right, and the one-shot `sessionStorage` guard avoids reload loops.
- **`Onboarding_Response`** kept WordPress-free and unit-tested across the real Frappe error shapes (nested / `_server_messages` / HTML-strip / malformed-JSON).
- Consistent `data-test-id`s and the `GODAM_E2E` short-circuit for Google sign-in.

### Verdict
**Approve with minor follow-ups.** None of the above blocks merge; #1 (enforce https) and #5 (accidental nudge dismissal) are the two I'd address before shipping.

*Note: the PR's own "known follow-ups" (forgot-password `api_key` scoping, WP-phpunit harness for the REST proxy + `should_overlay_onboarding()`, jest for the screens) are acknowledged and not re-raised here.*

**Review by KMchaudhary — APPROVED** [review]
As self-tested by developer, and resolve the claude code feedback, approving the PR
